### ISSUE TITLE
False negative for `unnecessary-dict-index-lookup` when inside function

### ISSUE THREAD

**Issue by clavedeluna** [PyCQA/pylint#7734]
### Bug description

I believe this code should emit `unnecessary-dict-index-lookup` message.

```
# test.py

FRUITS = {"apple": 1, "orange": 10, "berry": 22}


def pick_fruit(name):
    for fruit_name, count in FRUITS.items():
        if fruit_name == name:
            print(FRUITS[name])  # should be  [unnecessary-dict-index-lookup]
```

if I instead take the code out of the function AND remove the if statement, it does work

```
# def pick_fruit(name):
for fruit_name, count in FRUITS.items():
        # if fruit_name == name:
    print(FRUITS[fruit_name])  # [unnecessary-dict-index-lookup] correctly emitted

```

### Configuration

```ini
n/a
```

### Command used

```shell
pylint test.py --enable=all --enable-all-extensions
```

### Pylint output

```shell
************* Module test
test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test.py:4:0: C0116: Missing function or method docstring (missing-function-docstring)
test.py:5:20: W0612: Unused variable 'count' (unused-variable)
```

### Expected behavior

************* Module test
test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test.py:4:0: C0116: Missing function or method docstring (missing-function-docstring)doc/data/messages/u/unnecessary-test.py:7:n: R1733: Unnecessary dictionary index lookup, use 'count' instead (unnecessary-dict-index-lookup)

### Pylint version

```shell
pylint 2.16.0-dev
astroid 2.12.12
Python 3.8.9 (default, Apr 13 2022, 08:48:07)
[Clang 13.1.6 (clang-1316.0.21.2.5)]
```

### OS / Environment

Mac

**Comment by clavedeluna** [comment]
It does look like all the test cases in `tests/functional/u/unnecessary/unnecessary_dict_index_lookup.py` are module-level, so we should add function and class context tests and support for that.

**Comment by nickdrozd** [comment]
I guess the checker would have to verify that the index variable has the same value as the loop variable. Is there a convenient way to do that? It seems like it could be difficult in general.
